WcaOnRails/app/jobs/submit_report_nag_job.rb

```diff
@@ -3,20 +3,46 @@
 class SubmitReportNagJob < SingletonApplicationJob
   queue_as :default
 
-  def nag_needed(competition)
+  def nag_needed?(competition)
     (competition.delegate_report.nag_sent_at || competition.end_date) <= 8.days.ago
   end
 
+  def reminder_needed?(competition)
+    return unless competition.end_date
+    return if competition.delegate_report.reminder_sent_at
+
+    deadline = 7.days.since(competition.end_date)
+    1.day.before(deadline).today?
+  end
+
   def perform
+    competitions.each do |competition|
```

Comment:
This is not a a comment on the code itself, but rather an alternative approach we could have used if we had like 10 things to check like this.
You could create a array looking like this:
```ruby
updates = [
  { cond: "reminder_needed?", action: "submit_report_reminder", field: "reminder_sent_at"},
  { cond: "nag_needed?", action: "submit_report_nag", field: "nag_sent_at" },
]
```
And then used it in something like that to DRY it up:
```ruby
competitions.each do |c|
  updates.each do |up|
    if c.send(up[:cond])
      CompetitionsMailer.send(up[:action], c).deliver_now
      competition.delegate_report.update_attribute(up[:field], Time.now)
    end
  end
end
```

It's mostly an FYI, your code is perfectly fine as-is :)

Reply:
You suggest generalizing the code by introducing the abstraction (mini data DSL). But any abstractions have a cost. It's harder to read, harder to maintain. It's even harder to extend if you didn't predict the further evolution of the code :)
Yes, we remove code duplication. But at what cost? I believe that it will bring more harm than use.

Reply:
Which is why I wrote "if we had like 10 things to check", not 2 ^^

Arguably it's harder to maintain 10 similar parts of code than maintaining the "template" used for these 10 parts.